### PR TITLE
GLSL: Support GL_KHR_shader_subgroup_arithmetic (WIP)

### PR DESCRIPTION
**Pull request by georgeouzou**
Support OpenGL workarounds for subgroup arithmetic group operations.

NVIDIA workarounds are based on  GL_NV_shader_thread_shuffle.
AMD workarounds will be based on GL_AMD_shader_ballot.

Related issue  #1479

### CODE REVIEW

File: spirv_glsl.cpp [hunk 1]
```diff
@@ -3992,6 +3993,168 @@ void CompilerGLSL::emit_output_variable_initializer(const SPIRVariable &var)
 	}
 }
 
+void CompilerGLSL::emit_subgroup_arithmetic_workaround(std::string func, Op op, GroupOperation group_op)
```
Comment (HansKristian-Work) [inline spirv_glsl.cpp:3996]: const reference for func seems more appropriate.

File: spirv_glsl.cpp [hunk 2]
```diff
@@ -3992,6 +3993,168 @@ void CompilerGLSL::emit_output_variable_initializer(const SPIRVariable &var)
...
+	if (op_is_addition)
+	{
+		op_symbol = "+=";
+	}
+	else if (op_is_multiplication)
+	{
+		op_symbol = "*=";
+	}
+
+	for (const TypeInfo& t : type_infos) {
```
Comment (HansKristian-Work) [inline spirv_glsl.cpp:4095]: Brace on next line. `const TypeInfo &t`.

**Review by HansKristian-Work — CHANGES_REQUESTED** [review]
Hard to review the logic without tests.

Could you add arithmetic tests as well to demonstrate the codegen?

`shaders-no-opt/comp/subgroups_basicvoteballot.vk.comp` and
`reference/shaders-no-opt/comp/subgroups_basicvoteballot.vk.comp` is a test for the existing emulation I think.

**Comment by HansKristian-Work** [comment]
CI fails with a build error due to warning:

```
 /home/runner/work/SPIRV-Cross/SPIRV-Cross/spirv_glsl.cpp: In static member function ‘static bool spirv_cross::CompilerGLSL::is_supported_subgroup_op_in_opengl(spv::Op, const uint32_t*)’:
/home/runner/work/SPIRV-Cross/SPIRV-Cross/spirv_glsl.cpp:7354:3: error: this statement may fall through [-Werror=implicit-fallthrough=]
 7354 |   }
      |   ^
/home/runner/work/SPIRV-Cross/SPIRV-Cross/spirv_glsl.cpp:7356:2: note: here
 7356 |  default:
      |  ^~~~~~~
```

**Review by HansKristian-Work — APPROVED** [review]
Thanks for the tests, now it's more obvious what it's doing. Seems okay.

**Comment by georgeouzou** [comment]
@HansKristian-Work  thanks, i will add the other operations too. (max, min, etc)

**Comment by HansKristian-Work** [comment]
You can follow up in another PR with more opcodes.

**Pull request merged**